### PR TITLE
DDF-3747 Adds Result Form Support to Search Forms 

### PR DESCRIPTION
**Pull request by bdeining**
#### What does this PR do?
This PR adds support for Query Result Forms to be selected inside of the Query Forms, and tailors the result views to only show attributes specified on the Query Result Form.  Also fixes a bug when there is no template specified in user preferences.

~~STILL TODO:  Tailor the Inspector View to only show Query Result Form attributes~~

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
@adimka @mackncheesiest 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@andrewkfiedler
@jlcsmith
#### How should this be tested? (List steps with links to updated documentation)
Build / Install / Set up `etc/forms` and verify result forms are shown and can be selected / Verify Table 
View changes based on Query Form
#### What are the relevant tickets?
[DDF-3747](https://codice.atlassian.net/browse/DDF-3747)
#### Screenshots (if appropriate)
screen shot 2018-04-10 at 4 15 27 pm
screen shot 2018-04-10 at 4 15 23 pm
screen shot 2018-04-11 at 4 37 08 pm

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+ /*global require*/
+ var $ = require('jquery');
```
Comment (adimka) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: Since this is a new file, `const` would be preferable.

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        resultTemplates = data;
+    }
+ });
+
+ module.exports = Backbone.Model.extend({
+    initialize: function (options) {
+        this.options = options;
+        resultTemplatePromise.then(() => {
+            if (!this.isDestroyed) {
+                var customResultTemplates = _.map(resultTemplates, function(resultForm) {
```
Comment (adimka) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: Since this is a new file, `const` would be preferable.

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js (hunk 1 above)
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: done

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js (hunk 2 above)
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: done

File: ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/row.view.js [hunk 3]
```diff
@@ -115,7 +121,7 @@ module.exports = Marionette.LayoutView.extend({
                     property: property,
                     value: value,
                     class: className,
-                    hidden: hiddenColumns.indexOf(property) >= 0 || properties.isHidden(property) || metacardDefinitions.isHiddenTypeExceptThumbnail(property)
+                    hidden: (resultTemplateAttributes.length > 0 && resultTemplateAttributes.indexOf(property) == -1) || hiddenColumns.indexOf(property) >= 0 || properties.isHidden(property) || metacardDefinitions.isHiddenTypeExceptThumbnail(property)
```
Comment (adimka) [inline ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/row.view.js]: ✏️ This is a very long line :(

File: ui/packages/catalog-ui-search/src/main/webapp/component/editor/editor.view.js [hunk 4]
```diff
@@ -196,19 +196,27 @@ define([
...
+            var selectedResultTemplate = store.getCurrentQuery().get('selectedResultTemplate');
+            var resultTemplateAttributes = [];
+            if (selectedResultTemplate && selectedResultTemplate.descriptors) {
+               resultTemplateAttributes = selectedResultTemplate.descriptors;
             }
-            if (this.getEditorActionsOptions().summary){
-                var userSummaryChoice = user.get('user').get('preferences').get('inspector-summaryShown');
-                if (userSummaryChoice.length > 0){
-                    return userSummaryChoice.indexOf(attribute) >= 0;
+            if (resultTemplateAttributes.length == 0 || resultTemplateAttributes.indexOf(attribute) >= 0) {
```
[the remark this answers is not in the document]
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/editor/editor.view.js]: I'm not sure that `.includes` is supported in IE11

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js [hunk 5]
```diff
@@ -48,17 +53,44 @@ define([
...
+            if (this.model.get('resultForms')) {
+                var detailLevelProperty = new Property({
+                    label: 'Detail Level',
+                    enum: this.model.get('resultForms'),
+                    id: 'Detail Level',
+                    showValidationIssues: false,
+                    value: ['All Fields']
+                });
+
+                //this.handleChangeDetailLevel(detailLevelProperty, ['All Fields']);
```
[the remark this answers is not in the document]
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: Removed

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js [hunk 6]
```diff
@@ -48,17 +53,44 @@ define([
...
+            }
             this.querySettings.currentView.turnOffEditing();
             this.queryAdvanced.currentView.turnOffEditing();
             this.edit();
         },
+        handleChangeDetailLevel: function(model, values) {
+            $.each(model.get('enum') , (function(index, value) {
+                if (values[0] === value.value) {
+                    this.model.set('selectedResultTemplate', value);
+                    console.log(value);
```
[the remark this answers is not in the document]
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: Removed

File: ui/packages/catalog-ui-search/src/main/webapp/component/editor/editor.view.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (adimka) [reply ui/packages/catalog-ui-search/src/main/webapp/component/editor/editor.view.js]: I'm pretty sure it is, and if it isn't, we have babel polyfill in weback, so that'll make sure it's supported regardless of browser support. 👍

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+ /*global define,window*/
+
+ define([
```
Comment (adimka) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: We should use require instead of define for new files.

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form-tab-container.view.js [hunk 8]
```diff
@@ -18,6 +18,7 @@
  var template = require('./search-form.collection.hbs');
  var SearchFormCollectionView = require('./search-form.collection.view');
  var CustomElements = require('js/CustomElements');
+ var ResultForm = require('component/result-form/result-form');
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form-tab-container.view.js]: I couldn't find this being used.  Is it here for side effects?

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.view.js [hunk 9]
```diff
@@ -58,15 +57,17 @@
...
-                    type: 'custom',
-                    title: this.model.get('name')
-                });
+
                 user.getQuerySettings().set({
                     type: 'custom',
                     template: this.model.toJSON()
                 });
 
+                this.options.queryModel.set({
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.view.js]: Did the old order cause issues (maybe eventing problems)?

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.view.js]: Yeah, the user preferences were accessed before they were set, causing the first click to fail.

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form-tab-container.view.js (hunk 8 above)
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form-tab-container.view.js]: Removed unused import

**Comment by bdeining** [comment]
@andrewkfiedler There hasn't been much discussion as to whether or not this is a front-end or back-end capability and this functionality hasn't been driven by bandwidth concerns.  Moving this to a backend capability may work for Intrigue -> SOLR, but it wouldn't limit any resulting information from Sources.  So, if it is moved to the backend, then every source would need to be updated (if possible), which doesn't seem easier.

**Comment by middlets719** [comment]
Doing Hero

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ * License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+ /*global define,window*/
```
Comment (mackncheesiest) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: ✏️ Pretty sure these comments are only to tell jshint about globally defined variables so it doesn't freak out if they're used without declaring them, and we don't use jshint anymore

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js [hunk 11]
```diff
@@ -48,17 +54,39 @@ define([
...
+                this.resultTemplate.currentView.turnOnEditing();
+            }
             this.querySettings.currentView.turnOffEditing();
             this.queryAdvanced.currentView.turnOffEditing();
             this.edit();
         },
+        handleChangeDetailLevel: function(model, values) {
+            $.each(model.get('enum') , (function(index, value) {
+                if (values[0] === value.value) {
+                    this.model.set('selectedResultTemplate', value);
```
Comment (mackncheesiest) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: ❗️ We're setting values on the query model here that aren't actually reflected in [Query.js](https://github.com/codice/ddf/blob/master/ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js#L74-L88)

File: ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/thead.view.js [hunk 12]
```diff
@@ -96,6 +97,14 @@ module.exports = Marionette.ItemView.extend({
         var hiddenColumns = user.get('user').get('preferences').get('columnHide');
         var availableAttributes = this.options.selectionInterface.getActiveSearchResultsAttributes();
 
+        var resultTemplateAttributes = [];
+        if (store.getCurrentQuery()) {
+            var selectedResultTemplate = store.getCurrentQuery().get('selectedResultTemplate');
```
Comment (mackncheesiest) [inline ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/thead.view.js]: ❗️ Similar comment about accessing a model property that isn't defined in [Query.js](https://github.com/codice/ddf/blob/master/ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js#L74-L88)

File: ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/row.view.js [hunk 13]
```diff
@@ -82,6 +82,15 @@ module.exports = Marionette.LayoutView.extend({
         var prefs = user.get('user').get('preferences');
         var preferredHeader = user.get('user').get('preferences').get('columnOrder');
         var hiddenColumns = user.get('user').get('preferences').get('columnHide');
+
+        var resultTemplateAttributes = [];
+        if (store.getCurrentQuery()) {
+            var selectedResultTemplate = store.getCurrentQuery().get('selectedResultTemplate');
```
Comment (mackncheesiest) [inline ui/packages/catalog-ui-search/src/main/webapp/component/visualization/table/row.view.js]: ❗️ Similar comment about accessing a model property that isn't defined in [Query.js](https://github.com/codice/ddf/blob/master/ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js#L74-L88)

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js (hunk 11 above)
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: @mackncheesiest @andrewkfiedler Does it actually need to be reflected in Query.js?  It would be undefined in most cases

Comment (mackncheesiest) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: That's the model that is used throughout the UI for queries. If we're setting or retrieving anything on that model, I would certainly like it to be _at least_ assigned a default value of undefined in Query.js even if only for documentation purposes for developers.

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.view.js]: We can add the default value of undefined for documentation purposes.

**Comment by andrewkfiedler** [comment]
@bdeining I'm fine doing emulating the functionality we want on the backend on the frontend for now.  If we do so though I think we should chop off the attributes the moment they get back from the endpoint in the frontend.  That will limit the number of areas we have to make changes in and stop us from running into weird conflict scenarios with user prefs (since they can hide and show attributes in each visualization).  

So basically we should update QueryResponse to have the parse method take care of lopping off the attributes we don't want.  Or we can update the CQL endpoint to lop off the attributes we don't want before sending back the response from a source to the UI.  Either or.

**Comment by bdeining** [comment]
@andrewkfiedler made this change

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js [hunk 14]
```diff
@@ -52,9 +52,16 @@ function handleResultFormFields(result, selectedResultTemplate) {
...
-            _.each(resultAttributes, function(index, value) {
-                newProperties[value] = result.metacard.properties[value];
+            newProperties.id = result.metacard.properties.id;
+            newProperties["metacard-type"] = result.metacard.properties["metacard-type"];
+            _.each(resultAttributes, function(value, index) {
+                if (result.metacard.properties[value]) {
+                    newProperties[value] = result.metacard.properties[value];
+                }
             });
+            console.log(result.metacard.properties);
```
Comment (Schachte) [inline ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: remove

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js [hunk 15]
```diff
@@ -52,9 +52,16 @@ function handleResultFormFields(result, selectedResultTemplate) {
...
+            console.log(newProperties);
```
Comment (Schachte) [inline ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: remove

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: removed

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js (hunk 14 above)
Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: removed

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+ define([
+     'jquery',
+     'backbone'
+ ], function ($, Backbone) {
+
+  let resultTemplates = [];
+  let resultTemplateProperties = [];
+  let resultTemplatePromise = $.ajax({
+     type: 'GET',
+     context: this,
```
Comment (Schachte) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: Wouldn't you only need this if you used the context of this in the callback?

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.js]: yeah, nice catch, removed

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js [hunk 17]
```diff
@@ -47,6 +47,18 @@ function humanizeResourceSize(result) {
     }
 }
 
+function handleResultFormFields(result, selectedResultTemplate) {
+    if (selectedResultTemplate) {
+        var resultAttributes = selectedResultTemplate.descriptors;
```
[the remark this answers is not in the document]
Comment (adimka) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: All new files should use ES6 syntax, as well as `require` instead of `define`.
Existing files are developer's choice, and keeping style consistent in the file is totally fine.

**Comment by Schachte** [comment]
Hero In Progress

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (adimka) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: The @andrewkfiedler has spoken!

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form-interactions/search-form-interactions.view.js [hunk 18]
```diff
@@ -13,14 +13,14 @@
...
 /*global define, window*/
-var wreqr = require('wreqr');
-var Marionette = require('marionette');
-var template = require('./search-form-interactions.hbs');
-var CustomElements = require('js/CustomElements');
-var user = require('component/singletons/user-instance');
-var LoadingView = require('component/loading/loading.view');
-var announcement = require('component/announcement');
-var ConfirmationView = require('component/confirmation/confirmation.view');
+let wreqr = require('wreqr');
```
Comment (Schachte) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form-interactions/search-form-interactions.view.js]: Not to leave unnecessary nitpicks, but believe imports should be `const` to remain consistent with other files following ES6 syntax for imports.

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form-interactions/search-form-interactions.view.js]: done

**Comment by jlcsmith** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js [hunk 19]
```diff
@@ -47,6 +47,24 @@ function humanizeResourceSize(result) {
...
 }
 
+function handleResultFormFields(result, selectedResultTemplate) {
+    if (selectedResultTemplate) {
+        let resultAttributes = selectedResultTemplate.descriptors;
+        if (resultAttributes.length > 0) {
+            let newProperties = {};
+            newProperties.id = result.metacard.properties.id;
+            newProperties.title = result.metacard.properties.title ? result.metacard.properties.title : '';
+            newProperties["metacard-type"] = result.metacard.properties["metacard-type"];
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: Do we need tags as well?

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: They somehow were already being included in the query response, but i suppose it wouldn't hurt to explicitly add them here

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/js/model/QueryResponse.js]: done

**Comment by bdeining** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs [hunk 20]
```diff
@@ -20,6 +20,9 @@
         </div>
         <div class="query-settings">
 
+        </div>
+        <div class="query-result-template">
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs]: Why are we only allowing result detail templates for custom search templates?  I thought we would put this in the  settings component so that any search could utilize the result detail templates.

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs]: This is the first time i've heard that.

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs]: While the ticket was written before it was decided that `Basic` and `Text` were considered Search Forms, i'll refactor the PR.

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs]: Are there any other refactors that need to be done in this PR?

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-custom/query-custom.hbs]: @andrewkfiedler Moved to result template selector to the search settings view.

**Comment by bdeining** [comment]
build now

**Comment by bdeining** [comment]
added source id

**Comment by bdeining** [comment]
Build now
